rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -238,129 +218,124 @@ int main(int argc, char **argv){
-    dp->allocate_first_stage_output(total_output_samples_1);
+    dp->allocate_output(total_output_samples_1);
 
     dp->initial_memcpy_callback();
 
+    auto last_filter_output = dp->get_last_filter_output_d();
     call_decimate<DecimationType::bandpass>(dp->get_rf_samples_p(),
-      dp->get_first_stage_output_p(), dp->get_first_stage_bp_filters_p(), first_stage_dm_rate,
-      samples_needed, filters.get_first_stage_lowpass_taps().size(),
+      last_filter_output, dp->get_bp_filters_p(), dm_rates[0],
+      samples_needed, complex_taps[0].size(),
       rx_freqs.size(), total_antennas, rx_rate, dp->get_frequencies_p(),
-      "First stage of decimation", dp->get_cuda_stream());
+      "Bandpass stage of decimation", dp->get_cuda_stream());
 
 
+    std::vector<uint32_t> samples_per_antenna(complex_taps.size());
+    std::vector<uint32_t> total_output_samples(complex_taps.size());
+
+    samples_per_antenna[0] = num_output_samples_per_antenna;
+    total_output_samples[0] = total_output_samples_1;
 
     // When decimating, we go from one set of samples for each antenna in the first stage
     // to multiple sets of reduced samples for each frequency in further stages. Output samples are
     // grouped by frequency with all samples for each antenna following each other
     // before samples of another frequency start. In the first stage need a filter for each
     // frequency, but in the next stages we only need one filter for all data sets.
-    dp->allocate_and_copy_second_stage_filter(filters.get_second_stage_lowpass_taps().data(),
-                                                filters.get_second_stage_lowpass_taps().size());
-
-    auto num_output_samples_per_antenna_2 = num_output_samples_per_antenna_1 / second_stage_dm_rate;
-    auto total_output_samples_2 = rx_freqs.size() * num_output_samples_per_antenna_2 *
-                                    total_antennas;
+    cuComplex* prev_output = last_filter_output;
+    for (uint32_t i=1; i<complex_taps.size(); i++) {
+      samples_per_antenna[i] = samples_per_antenna[i-1]/dm_rates[i];
+      total_output_samples[i] = rx_freqs.size() * samples_per_antenna[i] * total_antennas;
 
-    dp->allocate_second_stage_output(total_output_samples_2);
+      dp->allocate_and_copy_lowpass_filter(complex_taps[i].data(), complex_taps[i].size());
+      dp->allocate_output(total_output_samples[i]);
 
-    // each antenna has a data set for each frequency after filtering.
-    auto samples_per_antenna_2 = total_output_samples_1/total_antennas/rx_freqs.size();
-    call_decimate<DecimationType::lowpass>(dp->get_first_stage_output_p(),
-      dp->get_second_stage_output_p(), dp->get_second_stage_filter_p(), second_stage_dm_rate,
-      samples_per_antenna_2, filters.get_second_stage_lowpass_taps().size(), rx_freqs.size(),
-      total_antennas, rx_rate, dp->get_frequencies_p(), "Second stage of decimation",
-      dp->get_cuda_stream());
+      auto last_lp_filter = dp->get_last_lowpass_filter_d();
```

Comment:
maybe the name of the variable could be allocated_filter and allocated_output instead of 'last'. ex:
allocated_lp_filter = dp->get_last_lowpass_filter_d(); 
Makes it easier to understand in the call_decimate call.

Reply:
changed.